<a href="https://colab.research.google.com/github/jproctor-rebecca/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment_RJProctor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [40]:
# YOUR WORK HERE
# chi-square = sum(observed-expected)^2/expected)
# critical chi^2 table https://people.richland.edu/james/lecture/m170/tbl-chi.html

import sys 
import numpy as np
import pandas as pd
from zipfile import ZipFile
from google.colab import files
from scipy import stats
from scipy.stats import chisquare
from scipy.stats import normaltest
from scipy.stats import kruskal
from scipy.stats import ttest_ind



In [43]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip


--2020-07-14 21:20:39--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip.1’

student.zip.1       100%[===================>]  20.00K  --.-KB/s    in 0.1s    

2020-07-14 21:20:40 (147 KB/s) - ‘student.zip.1’ saved [20478/20478]



In [50]:
!unzip student.zip
#ctrlF to find on page 

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [51]:
df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

In [52]:
df.describe(exclude='number')

school  sex address famsize  ... nursery higher internet romantic
count     395  395     395     395  ...     395    395      395      395
unique      2    2       2       2  ...       2      2        2        2
top        GP    F       U     GT3  ...     yes    yes      yes       no
freq      349  208     307     281  ...     314    375      329      263

[4 rows x 17 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [83]:
# 1st pairs of variables that are dependent
failures_absences_contingency = pd.crosstab(df['failures'], df['absences'], margins=True)

failures_absences_contingency

absences    0  1   2  3   4  5   6  7   8  ...  26  28  30  38  40  54  56  75  All
failures                                   ...                                     
0          89  3  58  6  47  4  25  6  17  ...   1   0   1   0   0   1   1   1  312
1          14  0   4  1   4  0   3  1   2  ...   0   1   0   1   1   0   0   0   50
2           5  0   1  0   2  0   2  0   2  ...   0   0   0   0   0   0   0   0   17
3           7  0   2  1   0  1   1  0   1  ...   0   0   0   0   0   0   0   0   16
All       115  3  65  8  53  5  31  7  22  ...   1   1   1   1   1   1   1   1  395

[5 rows x 35 columns]

In [84]:
chi2, p_value, dof, expected = stats.chi2_contingency(failures_absences_contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 122.18899006084001
p value 0.7958192721891358
degrees of freedom 136
expected frequencies table 
 [[9.08354430e+01 2.36962025e+00 5.13417722e+01 6.31898734e+00
  4.18632911e+01 3.94936709e+00 2.44860759e+01 5.52911392e+00
  1.73772152e+01 2.36962025e+00 1.34278481e+01 2.36962025e+00
  9.47848101e+00 2.36962025e+00 9.47848101e+00 2.36962025e+00
  5.52911392e+00 7.89873418e-01 3.94936709e+00 7.89873418e-01
  3.15949367e+00 7.89873418e-01 2.36962025e+00 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01 3.12000000e+02]
 [1.45569620e+01 3.79746835e-01 8.22784810e+00 1.01265823e+00
  6.70886076e+00 6.32911392e-01 3.92405063e+00 8.86075949e-01
  2.78481013e+00 3.79746835e-01 2.15189873e+00 3.79746835e-01
  1.51898734e+00 3.79746835e-01 1.51898734e+00 3.79746835e-01
  8.86075949e-01 1.26582278e-01 6.32911392e-01 1.26582278e-01
  5.06329114e-01 1.26582278e-01 3

1) Null Hypothesis: Independent

2) Alternative Hypothesis: Dependent

3) Confidence Level: 95%

Conclusion:

Due to a chi^2 statistic of 122.188 and a p-value of 0.795, we (***reject***/fail to reject) the null hypothesis that absences and failures are independent.

In [85]:
# 2nd pair of variables that are dependent
failures_internet_contingency = pd.crosstab(df['failures'], df['internet'], margins=True)

failures_internet_contingency

internet  no  yes  All
failures              
0         47  265  312
1         13   37   50
2          2   15   17
3          4   12   16
All       66  329  395

In [86]:
chi2, p_value, dof, expected = stats.chi2_contingency(failures_internet_contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 4.796825454131461
p value 0.7790546076737513
degrees of freedom 8
expected frequencies table 
 [[ 52.13164557 259.86835443 312.        ]
 [  8.35443038  41.64556962  50.        ]
 [  2.84050633  14.15949367  17.        ]
 [  2.67341772  13.32658228  16.        ]
 [ 66.         329.         395.        ]]


1) Null Hypothesis: Independent

2) Alternative Hypothesis: Dependent

3) Confidence Level: 95%

Conclusion:

Due to a chi^2 statistic of 4.796 and a p-value of 0.779, we (***reject***/fail to reject) the null hypothesis that internet access and failures are independent.

In [89]:
# 1st pair of variables that are independent
failures_guardian_contingency = pd.crosstab(df['failures'], df['guardian'], margins=True)

failures_guardian_contingency

guardian  father  mother  other  All
failures                            
0             71     230     11  312
1             14      25     11   50
2              4       6      7   17
3              1      12      3   16
All           90     273     32  395

In [90]:
chi2, p_value, dof, expected = stats.chi2_contingency(failures_guardian_contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 54.19591524000708
p value 2.5210642499921247e-07
degrees of freedom 12
expected frequencies table 
 [[ 71.08860759 215.63544304  25.27594937 312.        ]
 [ 11.39240506  34.55696203   4.05063291  50.        ]
 [  3.87341772  11.74936709   1.37721519  17.        ]
 [  3.64556962  11.05822785   1.29620253  16.        ]
 [ 90.         273.          32.         395.        ]]


1) Null Hypothesis: Independent

2) Alternative Hypothesis: Dependent

3) Confidence Level: 95%

Conclusion:

Due to a chi^2 statistic of 54.196 and a p-value of ~0, we (***reject***/fail to reject) the null hypothesis that failures and guardianship are independent.

In [91]:
# 2nd pair of variables that are independent
failures_paid_contingency = pd.crosstab(df['failures'], df['paid'], margins=True)

failures_paid_contingency

paid       no  yes  All
failures               
0         153  159  312
1          37   13   50
2          10    7   17
3          14    2   16
All       214  181  395

In [92]:
chi2, p_value, dof, expected = stats.chi2_contingency(failures_paid_contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 18.537217658927684
p value 0.017540454079544205
degrees of freedom 8
expected frequencies table 
 [[169.03291139 142.96708861 312.        ]
 [ 27.08860759  22.91139241  50.        ]
 [  9.21012658   7.78987342  17.        ]
 [  8.66835443   7.33164557  16.        ]
 [214.         181.         395.        ]]


1) Null Hypothesis: Independent

2) Alternative Hypothesis: Dependent

3) Confidence Level: 95%

Conclusion:

Due to a chi^2 statistic of 18.537 and a p-value of 0.0175, we (***reject***/fail to reject) the null hypothesis that internet and sex are independent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [53]:
# dependent by hand - internet & failures
contingency = pd.crosstab(df['failures'], df['internet'])

contingency

internet  no  yes
failures         
0         47  265
1         13   37
2          2   15
3          4   12

In [54]:
contingency.values

array([[ 47, 265],
       [ 13,  37],
       [  2,  15],
       [  4,  12]])

In [55]:
contingency = pd.crosstab(df['failures'], df['internet'], margins=True)

contingency

internet  no  yes  All
failures              
0         47  265  312
1         13   37   50
2          2   15   17
3          4   12   16
All       66  329  395

In [56]:
failures_no_internet = 312*66/395
failures_no_internet

52.131645569620254

In [57]:
row_totals = df['internet'].value_counts().sort_index().values

row_totals

array([ 66, 329])

In [62]:
col_totals = df['failures'].value_counts().sort_index().values

col_totals

array([312,  50,  17,  16])

In [71]:
sample_size = df.shape[0]

sample_size

395

In [74]:
expected = np.array([[0.0,0.0],
                     [0.0,0.0],
                     [0.0,0.0],
                     [0.0,0.0]])

for i , col in enumerate(col_totals):
  for j, row in enumerate(row_totals):
    expected_value = (row*col) / sample_size
    expected[i][j] = expected_value
    # i place holder for row position, python autofills
    # j place holder for col position, python autofills

expected

array([[ 52.13164557, 259.86835443],
       [  8.35443038,  41.64556962],
       [  2.84050633,  14.15949367],
       [  2.67341772,  13.32658228]])

In [ ]:
observed = np.array([[0.0,0.0],
                     [0.0,0.0],
                     [0.0,0.0],
                     [0.0,0.0]])

for k, col in enumerate( ):
  for l, row in enumerate( ):
    observed_value = (row*col) / sample_size
    observed[k][l] = observed_value

observed

### incomplete - could not figure out where to populate data from ###

In [ ]:

chi2 = ((observed - expected)**2 / expected).sum().sum()

chi2

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)